Sources/ShapedArray/Basic.swift

```diff
@@ -15,8 +15,21 @@
 
 extension ShapedArray {
     public func reshaped(to newShape: [Int]) -> ShapedArray {
-        precondition(newShape.reduce(1, *) == self.scalarCount, "Cannot reshape to shape \(newShape) because it has a different number of elements than the original shape \(self.shape).")
-        return .init(shape: newShape, scalars: self.scalars)
+        var shape = [Int]()
+        if newShape.contains(-1) {
+            let newShapeScalarCount = newShape.reduce(1, *) * -1
+            let remDim = self.scalarCount / newShapeScalarCount
+            for ele in newShape {
+                if ele == -1 { shape.append(remDim); continue }
+                shape.append(ele)
+            }
+        } else {
+            shape = newShape
+        }
```

Comment:
```suggestion
        let shape = {
            if newShape.contains(-1) {
                let newShapeScalarCount = newShape.reduce(1, *) * -1
                let remainingDim = self.scalarCount / newShapeScalarCount
                return newShape.map {
                    if $0 == -1 {
                        return remainingDim
                    }
                    return $0
                }
            } else {
                return newShape
            }
        }()
```
couple of points, prefer fuller names for variables, I honestly was confused by remDim at first, but figured it out based on context. prefer functional combinators like `map` over for-loops where sensible (here it's a pretty straight forward map).

i can go either way on assigning `var shape` vs executing a closure to `let shape` here, i think the closure is nicer since there's no mutability so i added that. the compiler should automatically inline in a release mode i believe. soon swift will have `if` expressions which will be nicer.

Reply:
I think these changes are good. Thank you for the comments!